src/core/MOM_porous_barriers.F90

```diff
@@ -0,0 +1,172 @@
+  type(verticalGrid_type),                    intent(in)  :: GV     !< The ocean's vertical grid structure.
+  type(unit_scale_type),                      intent(in)  :: US     !< A dimensional unit scaling type
+  real, dimension(SZI_(G),SZJ_(G),SZK_(G)),   intent(in)  :: h      !< Layer thicknesses [H ~> m or kg m-2]
+  type(thermo_var_ptrs),                      intent(in)  :: tv     !< A structure pointing to various
+                                                                    !! thermodynamic variables.
+  real, dimension(SZI_(G),SZJ_(G),SZK_(G)+1), intent(out) :: eta    !< layer interface heights
+                                                                    !! [Z ~> m] or 1/eta_to_m m).
+  real, dimension(SZI_(G),SZJ_(G)), optional, intent(in)  :: eta_bt !< optional barotropic
+             !! variable that gives the "correct" free surface height (Boussinesq) or total water
+             !! column mass per unit area (non-Boussinesq).  This is used to dilate the layer.
+             !! thicknesses when calculating interfaceheights [H ~> m or kg m-2].
+  integer,                          optional, intent(in)  :: halo_size !< width of halo points on
+                                                                       !! which to calculate eta.
+
+  real,                             optional, intent(in)  :: eta_to_m  !< The conversion factor from
+             !! the units of eta to m; by default this is US%Z_to_m.
+  type(porous_barrier_ptrs),           intent(inout) :: pbv  !< porous barrier fractional cell metrics
+
+  !local variables
+  integer ii, i, j, k, nk, isd, ied, jsd, jed, IsdB, IedB, JsdB, JedB
+  real w_layer, & ! fractional open width of layer interface [nondim]
+       A_layer, & ! integral of fractional open width from bottom to current layer[nondim]
+       A_layer_prev, & ! integral of fractional open width from bottom to previous layer [nondim]
+       eta_s, & ! layer height used for fit [Z ~> m]
+       eta_prev ! interface height of previous layer [Z ~> m]
+  isd = G%isd; ied = G%ied; jsd = G%jsd; jed = G%jed
+  IsdB = G%IsdB; IedB = G%IedB; JsdB = G%JsdB; JedB = G%JedB
+
+  !eta is zero at surface and decreases downward
+
+  nk = SZK_(G)
+
+  !currently no treatment for using optional find_eta arguments if present
+  call find_eta(h, tv, G, GV, US, eta)
+
+  do j=jsd,jed; do I=IsdB,IedB
+    if (G%porous_DavgU(I,j) < 0.) then
+      do K = nk+1,1,-1
+        eta_s = max(US%Z_to_m*eta(I,j,K), US%Z_to_m*eta(I+1,j,K)) !take shallower layer height
+        !eta_s = 0.5 * (US%Z_to_m*eta(I,j,K) + US%Z_to_m*eta(I+1,j,K)) !take arithmetic mean
+        if (eta_s <= G%porous_DminU(I,j)) then
+          pbv%por_layer_widthU(I,j,K) = 0.0
+          A_layer_prev = 0.0
+          if (K < nk+1) then
+            pbv%por_face_areaU(I,j,k) = 0.0; endif
+        else
+          call calc_por_layer(US%Z_to_m*(G%porous_DminU(I,j)-G%Z_ref), &
```

Comment:
The first 4 arguments to calc_por_layer() both here and in the v-face loop below should be left in units of [Z ~> m] and use the same reference depth GV%Z_ref, which will simplify the code here and enable the dimensional and reference-height consistency testing of the code inside of calc_por_layer().